File: backend/test/java/de/griefed/serverpackcreator/utilities/common/StringUtilitiesTest.java
package de.griefed.serverpackcreator.utilities.common;

import de.griefed.serverpackcreator.Dependencies;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;
import org.junit.jupiter.api.Assertions;
import org.junit.jupiter.api.Test;

public class StringUtilitiesTest {

  StringUtilities stringUtilities;

  StringUtilitiesTest() {
    stringUtilities = Dependencies.getInstance().UTILITIES().StringUtils();
  }

  @Test
  void buildStringTest() {
    List<String> args =
        new ArrayList<>(Arrays.asList("config", "mods", "scripts", "seeds", "defaultconfigs"));
    String result = stringUtilities.buildString(args.toString());
    Assertions.assertEquals(args.toString(), String.format("[%s]", result));
  }
}

File: serverpackcreator-api/src/test/kotlin/de/griefed/serverpackcreator/api/common/StringUtilitiesTest.kt
package de.griefed.serverpackcreator.api.common

import de.griefed.serverpackcreator.api.utilities.common.StringUtilities
import org.junit.jupiter.api.Assertions
import org.junit.jupiter.api.Test

class StringUtilitiesTest internal constructor() {
    @Test
    fun buildStringTest() {
        val args: List<String> = listOf("config", "mods", "scripts", "seeds", "defaultconfigs")
        val result: String = StringUtilities.buildString(args.toString())
        Assertions.assertEquals(args.toString(), "[%s]".format(result))
    }

    /**
     * A path free of forbidden characters is considered valid.
     */
    @Test
    fun checkForInvalidPathCharactersAcceptsCleanPath() {
        Assertions.assertTrue(StringUtilities.checkForInvalidPathCharacters("some/clean/path"))
        Assertions.assertTrue(StringUtilities.checkForInvalidPathCharacters(""))
    }

    /**
     * The presence of even a single forbidden character makes a path invalid. (Regression test for
     * the former OR-of-negations logic, which only flagged a path that contained *every* forbidden
     * character at once.)
     */
    @Test
    fun checkForInvalidPathCharactersRejectsAnySingleForbiddenCharacter() {
        for (forbidden in listOf("<", ">", ":", "\"", "|", "?", "*", "#", "%", "&", "{", "}", "$", "!", "@", "`", "´", "=")) {
            Assertions.assertFalse(
                StringUtilities.checkForInvalidPathCharacters("path${forbidden}name"),
                "A path containing '$forbidden' must be rejected"
            )
        }
    }
}